src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy

```diff
@@ -44,27 +46,72 @@ class VersionResolver {
         GitRepository repository = (GitRepository) this.repository
         Pattern releaseTagPattern = ~/^${tagProperties.prefix}.*/
         Pattern nextVersionTagPattern = ~/.*${nextVersionProperties.suffix}$/
-
-        TagsOnCommit tags = repository.latestTags(releaseTagPattern)
-        TagsOnCommit previousTags = tags;
-
-        Map currentVersionInfo = versionFromTags(tags.tags, false, nextVersionTagPattern, versionFactory)
+				
+        Map currentVersionInfo = null
+				Map previousVersionInfo = null
+				boolean isHead = false
+				
+				if (versionFactory.versionProperties.useHighestVersion) {
+					LinkedHashMap<String, List<String>> allTaggedCommits = repository.allTaggedCommits(releaseTagPattern, null, true)
+					currentVersionInfo = versionFromTaggedCommits(allTaggedCommits, false, nextVersionTagPattern, versionFactory)
+					String commitId = currentVersionInfo.commit
+					isHead = repository.isHeadCommit(commitId)
+					previousVersionInfo = versionFromTaggedCommits(allTaggedCommits, true, nextVersionTagPattern, versionFactory)
+				} 
+				else {
+					TagsOnCommit tags = repository.latestTags(releaseTagPattern)
+					TagsOnCommit previousTags = tags
+					isHead = tags.isHead
+					currentVersionInfo = versionFromTags(tags.tags, false, nextVersionTagPattern, versionFactory)
+					while (previousTags.hasOnlyMatching(nextVersionTagPattern)) {
+						previousTags = repository.latestTags(releaseTagPattern, previousTags.commitId)
+					}
+					previousVersionInfo = versionFromTags(previousTags.tags, true, nextVersionTagPattern, versionFactory)
+				}
+				
         Version currentVersion = currentVersionInfo.version
-
-
-        while (previousTags.hasOnlyMatching(nextVersionTagPattern)) {
-            previousTags = repository.latestTags(releaseTagPattern, previousTags.commitId)
-        }
-        Version previousVersion = versionFromTags(previousTags.tags, true, nextVersionTagPattern, versionFactory).version
+				Version previousVersion = previousVersionInfo.version
 
         return [
                 current         : currentVersion,
                 previous        : previousVersion,
-                onReleaseTag    : tags.isHead && !currentVersionInfo.isNextVersion,
+                onReleaseTag    : isHead && !currentVersionInfo.isNextVersion,
                 onNextVersionTag: currentVersionInfo.isNextVersion,
                 noTagsFound     : currentVersionInfo.noTagsFound
         ]
     }
+		
+		private Map versionFromTaggedCommits(LinkedHashMap<String, List<String>> taggedCommits,  boolean ignoreNextVersionTags, Pattern nextVersionTagPattern, VersionFactory versionFactory) {
```

Comment:
please break arguments declaration into multiple lines just like in `versionFromTags` method